### ISSUE TITLE
FreeCompany ID in Characters' JSON

### ISSUE THREAD

**Issue by Simbiat** [xivdb/api#1]
Hi!

You may remember me as the guy, who used your older PHP parser to create http://simbiat.ru/fftracker/
Due to LodeStone update it does not work now and I wanted to rely on XIVDB api now instead, where possible, but noticed, that characters' JSON do not have FreeCompany ID, which is a showstopper for me. Let alone that there is no JSON for FCs, but I am fine with dealing with it myself, if there is no alternative.
Is it planned to add FCID to character's? It would be very useful.

Thank you.

**Comment by viion** [comment]
Hi,

I will be, I recently did all the parsing for this and released my scripts, you could use that if you want to swap out the old PHP stuff, it's here: https://github.com/viion/lodestone-php

You'd use FC routes for parsing

```php
// FC Data
$url = sprintf(\Lodestone\Modules\Routes::LODESTONE_FREECOMPANY_URL, <id>);
$parser = new Lodestone\Parser\FreeCompany();
$data = $parser->url($url)->parse();

// FC members
$url = sprintf(\Lodestone\Modules\Routes::LODESTONE_FREECOMPANY_MEMBERS_URL, <id>);
$parser = new Lodestone\Parser\FreeCompany();
$data = $parser->url($url)->parse();
```

**Comment by Simbiat** [comment]
Oh, nice, I thought PHP version was abandoned. That definitely simplifies my migration now :D
Thanks you.

**Comment by Simbiat** [comment]
This is, probably a stupid question, but... I am supposed to require all of the php files in /src right? 'Cause when I do I still get "Class 'Lodestone\\Modules\\Routes' not found" error.
I am including them like this:
```
foreach (glob("./backend/lodestone-php-master/src/Lodestone/Dom/*.php") as $filename) {
    require_once $filename;
}
```

**Comment by viion** [comment]
Hello, it's intended to be used with composer, I don't provide an autoloader.

Have you used composer before?

**Comment by Simbiat** [comment]
Not really... In previous version there was an autoloader, and I used it with changing the folder to a proper one on my side. But in this case there are 3 folders, so... Not sure how to go about with it. Guess I'll try using ifs in autoloader to determine folder (= maybe that will help (=

**Comment by viion** [comment]
Making your own autoloader won't really work, the parser requires other modules https://packagist.org/packages/symfony/css-selector

Ideally you should use composer, nearly all the PHP world uses it so it's quite vital these days.

It supports all platforms: https://getcomposer.org/download/

Once you have it downloaded you would run (in command prompt or terminal): `composer require viion/lodestone-php`

If you just have the `composer.phar` file then you'd do `php composer.phar require viion/lodestone-php`

Once you have that you should see a `/vendor/` folder popup, it will have the lodestone-api in there and a bunch of other stuff.

Then you use composers autoload `require __DIR__.'/../vendor/autoload.php';`

View the `tests/cli.php` for an example.

**Comment by Simbiat** [comment]
I read on composer and SR0 or whatever and then careful placement and just a few lines solved it for me :D 
```php
class AutoLoad {
	function __construct() {
		spl_autoload_register(array(__CLASS__, 'autoload'));
	}
	public function autoload($classname) {
		$phpfile = __DIR__ .'/'. strtolower($classname) .'.php';
		require $phpfile;
	}
}
(new AutoLoad());
```
But when I run
```php
$parser = new Lodestone\Parser\FreeCompany();
$data = $parser->url(sprintf(\Lodestone\Modules\Routes::LODESTONE_FREECOMPANY_MEMBERS_URL, "9234631035923213559"))->parse();
```
I get these errors:

> [Fri Apr 14 18:45:31.248034 2017] [:error] [pid 11412:tid 2024] [client 192.168.1.5:50271] PHP Notice:  Trying to get property of non-object in root\\Lodestone\\Parser\\FreeCompany.php on line 84
> [Fri Apr 14 18:45:31.250039 2017] [:error] [pid 11412:tid 2024] [client 192.168.1.5:50271] PHP Notice:  Trying to get property of non-object in root\\Lodestone\\Parser\\ParserHelper.php on line 257
> [Fri Apr 14 18:45:31.250039 2017] [:error] [pid 11412:tid 2024] [client 192.168.1.5:50271] PHP Notice:  Undefined offset: 2 in root\\Lodestone\\Parser\\ParserHelper.php on line 258
> [Fri Apr 14 18:45:31.252057 2017] [:error] [pid 11412:tid 2024] [client 192.168.1.5:50271] PHP Notice:  Trying to get property of non-object in root\\Lodestone\\Parser\\FreeCompany.php on line 93
> [Fri Apr 14 18:45:31.254050 2017] [:error] [pid 11412:tid 2024] [client 192.168.1.5:50271] PHP Notice:  Trying to get property of non-object in root\\Lodestone\\Parser\\FreeCompany.php on line 98
> [Fri Apr 14 18:45:31.256055 2017] [:error] [pid 11412:tid 2024] [client 192.168.1.5:50271] PHP Notice:  Trying to get property of non-object in root\\Lodestone\\Parser\\FreeCompany.php on line 103
> [Fri Apr 14 18:45:31.257059 2017] [:error] [pid 11412:tid 2024] [client 192.168.1.5:50271] PHP Notice:  Trying to get property of non-object in root\\Lodestone\\Parser\\FreeCompany.php on line 105
> [Fri Apr 14 18:45:31.259064 2017] [:error] [pid 11412:tid 2024] [client 192.168.1.5:50271] PHP Notice:  Trying to get property of non-object in root\\Lodestone\\Parser\\FreeCompany.php on line 114
> [Fri Apr 14 18:45:31.269091 2017] [:error] [pid 11412:tid 2024] [client 192.168.1.5:50271] PHP Fatal error:  Uncaught Error: Call to a member function find() on boolean in root\\Lodestone\\Parser\\FreeCompany.php:145\nStack trace:\n#0 root\\Lodestone\\Parser\\FreeCompany.php(35): Lodestone\\Parser\\FreeCompany->parseFocus()\n#1 root\\test.php(8): Lodestone\\Parser\\FreeCompany->parse()\n#2 {main}\n  thrown in root\\Lodestone\\Parser\\FreeCompany.php on line 145

It works fine if I use LODESTONE_FREECOMPANY_URL route

**Comment by viion** [comment]
```php
$parser = new Lodestone\Parser\FreeCompanyMembers();
```

Use the FreeCompanyMembers parser.

**Comment by Simbiat** [comment]
Of course I downloaded it :D That was an obvious thing to do :D 
Thanks for the help. Checked everything and will now be able to restore my site functionality. Not like there are a lot of people who were using it beside me, though :D

**Comment by Simbiat** [comment]
Oh, one more thing, though: do I understand correct, that if Lodestone is unavailable, parser will return false?

**Comment by viion** [comment]
Yes you will get false if lodestone down or the FC was deleted, I need to put in maintaint check but I can only get the css when there is actually maintenance lol
